File: src/projects/AHF/AHF_1d_1.c
/* AHF_1d_1.c */
/* Copyright (C) 1998 Bernd Bruegmann, 27.5.2009 */
/* Produced with Mathematica */

#include "bam.h"
#include "AHF.h"

#define Power(x,y) (pow((double) (x), (double) (y)))
#define Log(x)     log((double) (x)))
#define pow2(x)    ((x)*(x))
#define pow2inv(x) (1.0/((x)*(x)))

#define Cal(x,y,z) ((x)?(y):(z))




void AHF_1d_1(tVarList *vars, tVarList *ahfvars)
{

tL *level = ahfvars->level;
double *xp1  = Ptr(level, "x");
double *xp2  = Ptr(level, "y");
double *xp3  = Ptr(level, "z");

double *g11 = vldataptr(vars, 0);
double *g12 = vldataptr(vars, 1);
double *g13 = vldataptr(vars, 2);
double *g22 = vldataptr(vars, 3);
double *g23 = vldataptr(vars, 4);
double *g33 = vldataptr(vars, 5);
double *A11 = vldataptr(vars, 6);
double *A12 = vldataptr(vars, 7);
double *A13 = vldataptr(vars, 8);
double *A22 = vldataptr(vars, 9);
double *A23 = vldataptr(vars, 10);
double *A33 = vldataptr(vars, 11);
double *K = vldataptr(vars, 12);
double *phi = vldataptr(vars, 13);
double *s1 = vldataptr(ahfvars, 0);
double *s2 = vldataptr(ahfvars, 1);
double *s3 = vldataptr(ahfvars, 2);
double *H = vldataptr(ahfvars, 3);


double dx = level->dx;
double dy = level->dy;
double dz = level->dz;
double oo2dx = 1/(2*dx), oo2dy = 1/(2*dy), oo2dz = 1/(2*dz);
double oodx2 = 1/(dx*dx), oody2 = 1/(dy*dy), oodz2 = 1/(dz*dz);
double oo4dxdy = 1/(4*dx*dy);
double oo4dxdz = 1/(4*dx*dz);
double oo4dydz = 1/(4*dy*dz);
long ccc;

double norm;
double S1;
double S2;
double S3;



forinner25(level,oo2dx,oo2dy,oo2dz) {


S1
=
xp1[ccc]
;

S2
=
xp2[ccc]
;

S3
=
xp3[ccc]
;

norm
=
Power(2.7182818284590452354,4.*phi[ccc])*
  (2.*(S1*(S2*g12[ccc] + S3*g13[ccc]) + S2*S3*g23[ccc]) + 
    g11[ccc]*pow2(S1) + g22[ccc]*pow2(S2) + g33[ccc]*pow2(S3))
;

s1[ccc]
=
xp1[ccc]/sqrt(norm)
;

s2[ccc]
=
xp2[ccc]/sqrt(norm)
;

s3[ccc]
=
xp3[ccc]/sqrt(norm)
;



} endforinner; /* loop i, j, k */

}  /* function */

/* AHF_1d_1.c */
/* nvars = 21, nauxs = 4, n* = 59,  n/ = 23,  n+ = 9, n = 91, O = 0 */

File: src/projects/AHF/AHF_1d_1.m
(* AHF_1d_1.m 
   mth 06/09 *)

(* compute right hand side of BSSN equations *)


(* variables *)
variables = { g[a,b],A[a,b],K,phi, s[a],H,
             xp[a]
            }
              


(* compute in this order *)
tocompute = {
    
    S[a] == xp[a],
    
    norm == Exp[4*phi]*g[a,b]*S[a]*S[b],
    
    s[a] == xp[a]/sqrt[norm]
    
}



(* symmetries *)

g[a_,b_] :=  g[b,a] /; !OrderedQ[{a,b}]
A[a_,b_] :=  A[b,a] /; !OrderedQ[{a,b}]





(************************************************************************)
(* information for C output *)

(* information about the function 
   the C file will be functionname.c 
*)
includeANDdefine[] := Module[{},

  pr["#include \"bam.h\"\n"];
  pr["#include \"AHF.h\"\n"];
  pr["\n"];
  pr["#define Power(x,y) (pow((double) (x), (double) (y)))\n"];
  pr["#define Log(x)     log((double) (x)))\n"];
  pr["#define pow2(x)    ((x)*(x))\n"];
  pr["#define pow2inv(x) (1.0/((x)*(x)))\n\n"];
  pr["#define Cal(x,y,z) ((x)?(y):(z))\n\n"];

  pr["\n\n\n"];
];

functionname = "AHF_1d_1";


functionarguments = "tVarList *vars, tVarList *ahfvars";


functionloop = 
  "forinner25(level,oo2dx,oo2dy,oo2dz) {";
endfunctionloop = "} endforinner; /* loop i, j, k */";







(* custom variable declaration
   we have to translate between the tensor names and the C variables  
   we may need special variables, for example for parameters
*)
variabledeclarations[] := Module[{},

  pr["tL *level = ahfvars->level;\n"];
  pr["double *xp1  = Ptr(level, \"x\");\n"];
  pr["double *xp2  = Ptr(level, \"y\");\n"];
  pr["double *xp3  = Ptr(level, \"z\");\n"];
  pr["\n"];
  
 
  prdecvl[{ g[a,b],A[a,b],K,phi}, "vars"];
  prdecvl[{ s[a],H}, "ahfvars"];
 
  
  pr["\n"];
];    




(************************************************************************)
(* now we are ready to go *)

<< "../../math/MathToC/TensorEquationsToC.m"
